### PR TITLE
Experiment with fixing the PIN continue button issue

### PR DESCRIPTION
**Pull request by bobsilverberg**
As per our discussion in IRC, @davehunt, I tried to implement a similar solution to the one we used for Login for the fact that we often fail to click on the PIN "Continue" or "Submit" button after entering the PIN. 

As you will see in the attached diff, I removed the previous code which did:
```Wait(marionette=self.marionette).until(lambda m: button.is_enabled())```

and replaced it with a sequence that waits for the element to be present, then displayed, then enabled, and then scrolls it into view.

When I did that, without the sleep at line 87, it *never* works. The tap *always* fails, whereas with the old code it only fails sometimes. With the added sleep it always seems to work.

This suggests to me that the old code was taking a bit longer to execute than the new code, which is why it worked sometimes. It was like a sleep. I am a bit surprised that the old code takes longer to execute than the new code, as it's essentially doing the same thing (it's doing even less, in fact), but I guess the way each method is implemented in Marionette is different.

So right now the only working solution I have is the `sleep`. Do you have any other suggestions for better things to do/wait for in this case?

To reproduce, run https://github.com/mozilla/marketplace-tests-gaia/blob/master/marketplacetests/tests/test_marketplace_create_confirm_pin.py

### CODE REVIEW

File: marketplacetests/payment/app.py [hunk 1]
```diff
@@ -80,8 +80,15 @@ def tap_cancel_pin(self):
         self.apps.switch_to_displayed_app()
 
     def tap_pin_continue(self):
-        button = self.marionette.find_element(*self._pin_continue_button_locator)
-        Wait(marionette=self.marionette).until(lambda m: button.is_enabled())
+        button = Wait(self.marionette).until(
+            expected.element_present(*self._pin_continue_button_locator))
+        Wait(self.marionette).until(expected.element_displayed(button))
+        Wait(self.marionette).until(expected.element_enabled(button))
+        time.sleep(1)
```
Comment (bobsilverberg) [inline marketplacetests/payment/app.py]: Without this sleep I never see the button successfully clicked.

**Comment by bobsilverberg** [comment]
@davehunt I did as we discussed and grabbed the page source at all three steps:
1. As soon as the `tap_pin_continue` is called (i.e., the first line of the method).
2. Immediately after the `Wait(self.marionette).until(expected.element_enabled(button))`.
3. After a `sleep(2)` after the wait for enabled.

Unfortunately the page source is identical in all three cases, so there doesn't appear to be anything about the markup that we can use to check that the element is ready to be interacted with.

If you have any other suggestions, they would be welcome.

**Comment by davehunt** [comment]
Ive been unable to replicate your issue without the sleep, but the keyboard is pushing the elements around, so it's possible we should wait for things to settle. Alternatively, we could use the on screen keyboard to enter the PIN. I have experimented a little with this, and the following patch works for me:

```patch
diff --git a/marketplacetests/payment/app.py b/marketplacetests/payment/app.py
index 99cac1a..41bdf05 100644
--- a/marketplacetests/payment/app.py
+++ b/marketplacetests/payment/app.py
@@ -56,16 +56,17 @@ class Payment(Base):
         return self.marionette.find_element(*self._pin_heading_locator).text
 
     def create_pin(self, pin):
-        self.wait_for_element_displayed(*self._pin_container_locator)
-        Wait(marionette=self.marionette).until(lambda m: 'Create' in self.pin_heading)
-        self.marionette.find_element(*self._pin_container_locator).send_keys(pin)
+        Wait(self.marionette).until(lambda m: 'Create' in self.pin_heading)
+        self.keyboard.send(pin)
+        self.switch_to_payment_frame()
         self.tap_pin_continue()
 
-        # Workaround click because Marionette makes the keyboard disappear
-        self.marionette.find_element(*self._pin_container_locator).click()
-
-        Wait(marionette=self.marionette).until(lambda m: 'Confirm' in self.pin_heading)
-        self.marionette.find_element(*self._pin_container_locator).send_keys(pin)
+        Wait(self.marionette).until(lambda m: 'Confirm' in self.pin_heading)
+        element = Wait(self.marionette).until(
+            expected.element_present(*self._pin_container_locator))
+        element.click()
+        self.keyboard.send(pin)
+        self.switch_to_payment_frame()
         self.tap_pin_continue()
 
     def enter_pin(self, pin):
@@ -84,11 +85,6 @@ class Payment(Base):
             expected.element_present(*self._pin_continue_button_locator))
         Wait(self.marionette).until(expected.element_displayed(button))
         Wait(self.marionette).until(expected.element_enabled(button))
-        time.sleep(1)
-        # This workaround is required for gaia v2.0, but can be removed in later versions
-        # as the bug has been fixed
-        # Bug 937053 - tap() method should calculate elementInView from the coordinates of the tap
-        self.marionette.execute_script('arguments[0].scrollIntoView(false);', [button])
         button.tap()
 
     def wait_for_buy_app_section_displayed(self):
```

**Comment by bobsilverberg** [comment]
Thanks @davehunt. I used your latest suggestions and it seems to have stabilized. I also included a refactor here (to extract `confirm_pin` into its own method) which I did in a later PR, just so we can have it all here now.

Another r? would be appreciated.

File: marketplacetests/payment/app.py [hunk 2]
```diff
@@ -56,22 +54,36 @@ def pin_heading(self):
...
 
-        Wait(marionette=self.marionette).until(lambda m: 'Confirm' in self.pin_heading)
-        self.marionette.find_element(*self._pin_container_locator).send_keys(pin)
+    def confirm_pin(self, pin):
+        Wait(self.marionette).until(lambda m: 'Confirm' in self.pin_heading)
+        element = Wait(self.marionette).until(
+            expected.element_present(*self._pin_container_locator))
+        Wait(self.marionette).until(
+            expected.element_displayed(element))
+        element.click()
```
Comment (davehunt) [inline marketplacetests/payment/app.py]: Nit: Add a comment here that the click is necessary to show the keyboard. A tap sadly does not work, and for some reason the keyboard does not show automatically as it does when it's being driven manually.

**Comment by davehunt** [comment]
Just one nit, and there's a bit of duplicate code here. I'm fine with this, but we should consider refactoring in the future as a low priority. Once the nit is addressed please feel free to merge this.

**Comment by bobsilverberg** [comment]
Thanks @davehunt. I addressed your nit, and agree that some refactoring is called for, but that will be addressed in a later PR after all the new tests have been added.

**Pull request merged**